File: src/test/java/pl/laskowski/marcin/model/matrix/ArrayMatrixTest.java
package pl.laskowski.marcin.model.matrix;

import org.junit.Test;

import static org.junit.Assert.*;

/**
 * Created by Marcin Laskowski.
 */

public class ArrayMatrixTest {

    @Test
    public void createEmptyMatrixAndCheckItsSize() {
        Matrix matrix = new ArrayMatrix(0, 0);

        assertEquals(0, matrix.sizeX());
        assertEquals(0, matrix.sizeY());
    }

    @Test
    public void createNonEmptyMatrixAndCheckItsSize() {
        Matrix<Object> matrix = new ArrayMatrix<>(4, 7);

        assertEquals(4, matrix.sizeX());
        assertEquals(7, matrix.sizeY());
    }

    @Test
    public void checkIfMatrixIsInitializedWithNulls() {
        Matrix<Object> matrix = new ArrayMatrix<>(2, 2);

        assertNull(matrix.get(0, 0));
        assertNull(matrix.get(1, 0));
        assertNull(matrix.get(0, 1));
        assertNull(matrix.get(1, 1));
    }

    @Test
    public void insertObjectIntoMatrixAndCheckItsThere() {
        Matrix<Object> matrix = new ArrayMatrix<>(3, 3);
        Object object = new Object();

        matrix.put(object, 1, 2);

        assertSame(object, matrix.get(1, 2));
    }

    @Test
    public void checkIfMatrixHandlesEmptyStateProperly() {
        Matrix<Object> matrix = new ArrayMatrix<>(3, 3);
        assertTrue(matrix.isEmpty());

        matrix.put(new Object(), 1, 1);
        assertFalse(matrix.isEmpty());

        matrix.put(null, 1, 1);
        assertTrue(matrix.isEmpty());
    }

    @Test
    public void checkIfClearingMatrixWorks() {
        Matrix<Object> matrix = new ArrayMatrix<>(3, 3);

        matrix.put(new Object(), 1, 1);
        matrix.clear();

        assertTrue(matrix.isEmpty());
    }

    @Test
    public void checkIfMatrixContainsGivenElement() {
        Matrix<Integer> matrix = new ArrayMatrix<>(2, 2);
        int testValue = 7;

        assertFalse(matrix.contains(testValue));

        matrix.put(testValue, 1, 1);
        assertTrue(matrix.contains(testValue));
    }

    @Test
    public void removeGivenElementFromMatrix() {
        Matrix<Integer> matrix = new ArrayMatrix<>(2, 2);
        int testValue = 7;

        matrix.put(testValue, 1, 1);
        matrix.remove(testValue);

        assertFalse(matrix.contains(testValue));
    }

    @Test
    public void checkIfGivenElementWasRemoved() {
        Matrix<Integer> matrix = new ArrayMatrix<>(2, 2);
        int testValue = 7;

        matrix.put(testValue, 1, 1);

        assertTrue(matrix.remove(testValue));
        assertFalse(matrix.remove(testValue));
    }

    @Test
    public void checkIfMatrixIsEqualsWhenSizeIsEquals() {
        Matrix<Object> smallMatrix = new ArrayMatrix<>(2, 2);
        Matrix<Object> otherSmallMatrix = new ArrayMatrix<>(2, 2);
        Matrix<Object> veryBigMatrix = new ArrayMatrix<>(1000, 1000);

        assertTrue(smallMatrix.equals(otherSmallMatrix));
        assertFalse(smallMatrix.equals(veryBigMatrix));

        assertTrue(smallMatrix.hashCode() == otherSmallMatrix.hashCode());
        assertFalse(smallMatrix.hashCode() == veryBigMatrix.hashCode());
    }

    @Test
    public void checkIfMatrixIsEqualsWhenDataIsEquals() {
        Matrix<String> twinMatrix1 = new ArrayMatrix<>(2, 2);
        twinMatrix1.put("twin", 0, 0);
        Matrix<String> twinMatrix2 = new ArrayMatrix<>(2, 2);
        twinMatrix2.put("twin", 0, 0);
        Matrix<String> differentMatrix1 = new ArrayMatrix<>(2, 2);
        differentMatrix1.put("not twin", 0, 0);
        Matrix<String> differentMatrix2 = new ArrayMatrix<>(2, 2);
        differentMatrix2.put("twin", 1, 1);

        assertTrue(twinMatrix1.equals(twinMatrix2));
        assertFalse(twinMatrix1.equals(differentMatrix1));
        assertFalse(twinMatrix1.equals(differentMatrix2));

        assertTrue(twinMatrix1.hashCode() == twinMatrix2.hashCode());
        assertFalse(twinMatrix1.hashCode() == differentMatrix1.hashCode());
        assertFalse(twinMatrix1.hashCode() == differentMatrix2.hashCode());
    }

}

File: app/src/test/java/pl/laskowski/marcin/model/matrix/ArrayMatrixTest.kt
package pl.laskowski.marcin.model.matrix

import org.junit.Assert.*
import org.junit.Test

/**
 * Created by Marcin Laskowski.
 */
class ArrayMatrixTest {

    @Test
    fun createEmptyMatrixAndCheckItsSize() {
        val matrix = ArrayMatrix<Any>(0, 0)
        assertEquals(0, matrix.sizeX())
        assertEquals(0, matrix.sizeY())
    }

    @Test
    fun createNonEmptyMatrixAndCheckItsSize() {
        val matrix = ArrayMatrix<Any>(4, 7)
        assertEquals(4, matrix.sizeX())
        assertEquals(7, matrix.sizeY())
    }

    @Test
    fun checkIfMatrixIsInitializedWithNulls() {
        val matrix = ArrayMatrix<Any>(2, 2)
        assertNull(matrix[0, 0])
        assertNull(matrix[1, 0])
        assertNull(matrix[0, 1])
        assertNull(matrix[1, 1])
    }

    @Test
    fun insertObjectIntoMatrixAndCheckItsThere() {
        val matrix = ArrayMatrix<Any>(3, 3)
        val any = Any()

        matrix.put(any, 1, 2)
        assertSame(any, matrix[1, 2])
    }

    @Test
    fun checkIfMatrixHandlesEmptyStateProperly() {
        val matrix = ArrayMatrix<Any>(3, 3)
        assertTrue(matrix.isEmpty)

        matrix.put(Any(), 1, 1)
        assertFalse(matrix.isEmpty)

        matrix.put(null, 1, 1)
        assertTrue(matrix.isEmpty)
    }

    @Test
    fun checkIfClearingMatrixWorks() {
        val matrix = ArrayMatrix<Any>(3, 3)

        matrix.put(Any(), 1, 1)
        matrix.clear()

        assertTrue(matrix.isEmpty)
    }

    @Test
    fun checkIfMatrixContainsGivenElement() {
        val matrix = ArrayMatrix<Int>(2, 2)
        val testValue = 7

        assertFalse(matrix.contains(testValue))

        matrix.put(testValue, 1, 1)

        assertTrue(matrix.contains(testValue))
    }

    @Test
    fun removeGivenElementFromMatrix() {
        val matrix = ArrayMatrix<Int>(2, 2)
        val testValue = 7

        matrix.put(testValue, 1, 1)
        matrix.remove(testValue)

        assertFalse(matrix.contains(testValue))
    }

    @Test
    fun checkIfGivenElementWasRemoved() {
        val matrix = ArrayMatrix<Int>(2, 2)
        val testValue = 7

        matrix.put(testValue, 1, 1)

        assertTrue(matrix.remove(testValue))
        assertFalse(matrix.remove(testValue))
    }

    @Test
    fun checkIfMatrixIsEqualsWhenSizeIsEquals() {
        val smallMatrix = ArrayMatrix<Any>(2, 2)
        val otherSmallMatrix = ArrayMatrix<Any>(2, 2)
        val veryBigMatrix = ArrayMatrix<Any>(1000, 1000)

        assertEquals(smallMatrix, otherSmallMatrix)
        assertNotEquals(smallMatrix, veryBigMatrix)

        assertEquals(smallMatrix.hashCode(), otherSmallMatrix.hashCode())
        assertNotEquals(smallMatrix.hashCode(), veryBigMatrix.hashCode())
    }

    @Test
    fun checkIfMatrixIsEqualsWhenDataIsEquals() {
        val twinMatrix1 = ArrayMatrix<String>(2, 2)
        twinMatrix1.put("twin", 0, 0)

        val twinMatrix2 = ArrayMatrix<String>(2, 2)
        twinMatrix2.put("twin", 0, 0)

        val differentMatrix1 = ArrayMatrix<String>(2, 2)
        differentMatrix1.put("not twin", 0, 0)

        val differentMatrix2 = ArrayMatrix<String>(2, 2)
        differentMatrix2.put("twin", 1, 1)

        assertEquals(twinMatrix1, twinMatrix2)
        assertNotEquals(twinMatrix1, differentMatrix1)
        assertNotEquals(twinMatrix1, differentMatrix2)

        assertEquals(twinMatrix1.hashCode(), twinMatrix2.hashCode())
        assertNotEquals(twinMatrix1.hashCode(), differentMatrix1.hashCode())
        assertNotEquals(twinMatrix1.hashCode(), differentMatrix2.hashCode())
    }
}